src/topology/perfect.lean

```diff
@@ -211,3 +218,342 @@ begin
 end
 
 end kernel
+end basic
+
+section scheme
+
+/-- A `β`-scheme on `α` is a function from `list β` to `set α`.
+We typically think of this as a "tree" of subsets of `α`, and use the appropriate terminology
+(branch, children, etc.).
+The usefulness of such a scheme is that a map `(ℕ → β) → α` can often be thought of as
+a sort of "limiting object" of a `β`-scheme on `α`. -/
+def set.scheme (β α : Type*) := list β → set α
+
+namespace set.scheme
+open list metric function
+open_locale classical
+
+instance inhabited {β α : Type*} : inhabited (set.scheme β α) :=
+begin
+  rw set.scheme,
+  apply_instance,
+end
+
+variables {β α : Type*} (A : set.scheme β α)
+
+/-- `res x n`, or the restriction of `x` to `n`,
+is the list of length `n` whose `m`-th entry is `x m`.-/
+def res (x : ℕ → β) : ℕ → list β
+  | 0            := nil
+  | (nat.succ n) := (res n).concat (x n)
+
+@[simp] lemma res_zero (x : ℕ → α) : res x 0 = @nil α := rfl
+@[simp] lemma res_succ (x : ℕ → α) (n : ℕ) : res x n.succ = (res x n).concat (x n) := rfl
+
+@[simp] lemma res_length (x : ℕ → α) (n : ℕ) : (res x n).length = n :=
+begin
+  induction n with n ih,
+  { refl },
+  simp[ih],
+end
+
+/-- The restrictions of `x` and `y` to `n` are equal if and only if `x m = y m` for all `m < n`.-/
+lemma res_eq_iff (x y : ℕ → α) (n : ℕ) : res x n = res y n ↔ ∀ m < n, x m = y m :=
+begin
+  split; intro h; induction n with n ih, { simp },
+  { intros m hm,
+    rw nat.lt_succ_iff_lt_or_eq at hm,
+    rw [← reverse_inj] at h,
+    simp only [res_succ, concat_eq_append, reverse_append, reverse_singleton,
+      singleton_append, reverse_inj] at h,
+    cases hm with hm hm,
+    { exact ih h.2 _ hm },
+    rw hm,
+    exact h.1, },
+  { simp },
+  simp only [res_succ],
+  rw [ih _, h _ (nat.lt_succ_self _)],
+  intros m hmn,
+  exact h m (hmn.trans (nat.lt_succ_self _)),
+end
+
+/-- Two infinite sequences are equal if and only if all their restrictions are.-/
+theorem eq_iff_res_eq (x y : ℕ → α) : (∀ n, res x n = res y n) ↔ x = y :=
+begin
+  split; intro h,
+  { ext n,
+    specialize h n.succ,
+    rw res_eq_iff at h,
+    exact h _ (nat.lt_succ_self _), },
+  rw h,
+  simp,
+end
+
+/-- `cylinder x n` is equal to the set of sequences `y` with the same restriction to `n` as `x`.-/
+theorem cylinder_eq_res (x : ℕ → α) (n : ℕ) : pi_nat.cylinder x n = {y | res y n = res x n} :=
+begin
+  ext y,
+  dsimp[pi_nat.cylinder],
+  rw res_eq_iff,
+end
+
+/-- From a `β`-scheme on `α` `A`, we define a partial function from `(ℕ → β)` to `α`
+which sends each infinite sequence `x` to an element of the intersection along the
+branch corresponding to `x`, if it exists.
+We call this the map induced by the scheme. -/
+noncomputable def map : Σ s : set (ℕ → β), s → α :=
```

Comment:
This is a pretty bad name IMO because this is not a map coming from a functor. Consider finding a more descriptive (ha!) name.

Reply:
I've changed it to `induced_map`. Not sure what else to call it. I don't know if there's a very standard name. Kechris calls it the associated map which is probably worse.